core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -536,26 +536,31 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
         return new Engine.Index(uid, doc, version, versionType, origin, startTime, autoGeneratedIdTimestamp, isRetry);
     }
 
-    public void index(Engine.Index index) {
+    public Engine.IndexResult index(final Engine.Index index) {
         ensureWriteAllowed(index);
         Engine engine = getEngine();
-        index(engine, index);
+        return index(engine, index);
     }
 
-    private void index(Engine engine, Engine.Index index) {
+    private Engine.IndexResult index(final Engine engine, final Engine.Index index) {
         active.set(true);
-        index = indexingOperationListeners.preIndex(index);
+        final Engine.IndexResult result;
+        indexingOperationListeners.preIndex(index);
         try {
             if (logger.isTraceEnabled()) {
                 logger.trace("index [{}][{}]{}", index.type(), index.id(), index.docs());
             }
-            engine.index(index);
-            index.endTime(System.nanoTime());
+            result = engine.index(index);
         } catch (Exception e) {
             indexingOperationListeners.postIndex(index, e);
             throw e;
         }
-        indexingOperationListeners.postIndex(index, index.isCreated());
+        if (result.hasFailure()) {
```

Comment:
I'm wondering if should make the distinction here and not just call postIndex(index,result) where the result is a failure